src/foam/dao/MongoDAO.java

```diff
@@ -0,0 +1,176 @@
+      MongoCredential credential = MongoCredential.createCredential(username, dbName, password.toCharArray());
+      mongoClient = new MongoClient(new ServerAddress(host, port), Arrays.asList(credential));
+    } else {
+      mongoClient = new MongoClient(host, port);
+    }
+
+    this.database = mongoClient.getDatabase(dbName);
+  }
+
+  public MongoDAO(String dbName, String username, String password) {
+    this("localhost", 27017, dbName, username, password);
+  }
+
+  public MongoDAO(String dbName) {
+    this(dbName, null, null);
+  }
+
+  private boolean isUserPassProvided(String username, String password) {
+    return ( ( username != null ) && ( ! username.isEmpty() ) ) &&
+            ( ( password != null ) && ( ! password.isEmpty() ) );
+  }
+
+  @Override
+  public Sink select_(X x, Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
+    if ( sink == null ) {
+      sink = new ListSink();
+    }
+
+    Sink         decorated = decorateSink_(sink, skip, limit, order, predicate);
+    Subscription sub       = new Subscription();
+
+    String collectionName = (String) x.get("collectionName");
+    String collectionClass = (String) x.get("collectionClass");
+    List props = getFObjectProperties(collectionClass, x);
+
+    if ( collectionName == null || collectionName.isEmpty() ) {
+      throw new IllegalArgumentException("Invalid collection name in context. Please provide a string name.");
+    } else if ( collectionClass == null || collectionClass.isEmpty() ) {
+      throw new IllegalArgumentException("Invalid collection class name in context. Please provide a string class name.");
+    }
+
+    MongoCollection<Document> collection = database.getCollection(collectionName);
+    MongoCursor<Document> cursor = collection.find().iterator();
+
+    try {
+      while ( cursor.hasNext() ) {
+        if ( sub.getDetached() ) break;
+
+        FObject obj = createFObject(collectionClass, props, cursor.next());
+
+        if ( ( predicate == null ) || predicate.f(obj) ) {
+          decorated.put(obj, sub);
+        }
+      }
+    } finally {
+      cursor.close();
+    }
+
+    decorated.eof();
+
+    return sink;
+  }
+
+  private List getFObjectProperties(String clsName, X x) {    
+    try {
+      Class cls = Class.forName(clsName);
+      FObject clsInstance = (FObject) x.create(cls);
+      List props = clsInstance.getClassInfo().getAxiomsByClass(foam.core.AbstractFObjectPropertyInfo.class);
+      
+      // Recursively handle nested FObjectProperties
+      for ( int i = 0 ; i < props.size() ; i++ ) {
+        props.addAll(getFObjectProperties(((PropertyInfo) props.get(i)).getPropertyType(), x));
+      }
+
+      return props;
+    } catch (Exception ex) {
+      throw new RuntimeException(clsName + " was not found.");
+    }
+  }
+
+  private FObject createFObject(String clsName, List props, Document d) {
+    JsonWriterSettings writerSettings = new JsonWriterSettings(JsonMode.SHELL, true);
```

Comment:
This method will be really really slow.  You're creating a lot of garbage, first by serializing the Mongo Document to JSON, and then appending a bunch more strings together to build a FOAM JSOn string, and then parsing that back to a FOAM object.

This is being called for every single object enumerated in a .select() call, meaning we're creating a lot of garbage in a tight loop, that will be bad for performance.


A better approach would be to create the FObject first with X.create(getOf()), and then set its properties individually as you enumerate the fields of the document.

You also may need to do a little bit of serialization/deserialization work by hand here for certain property values that MongoDB can't store out of box.